processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java

```diff
@@ -917,7 +917,11 @@ public PropertyMapping build() {
             );
 
             Assignment assignment = null;
-            if ( !targetType.isEnumType() ) {
+            if ( targetType.equals( sourceType ) ) { // No need to lookup a mapping.
```

Comment:
I'm not sure here whether there are no side effects. I realise this is a constant. However, zooming out what do you do when the user explicitly wants to convert a constant or default value to a String? 

In other words: do we see a use case were a constant String is converted into another String? At first glance this sounds meaningless. But we do have context objects. Although, perhaps far sought..

So:
```java
@Mapper
 public interface Issue2636Mapper {
     @Mapping( source = "folder.ancestor.id", target = "parent", defaultValue = "#", qualifiedByName = "uuidToString" )
     DirectoryNode convert(Folder folder, @Context State someState);

     @Named( "uuidToString" )
     default String uuidToString(UUID id, @Context State someState) {
         // do something with state and
         return someState.doSomethingWithId(id);
     }

     @Named( "uuidToString" )
     default String defaultToString(String id, @Context State someState) {
         // do something with state and
         return someState.doSomethingWithId(id);
     }

 }
```
BTW: I'm curious to know it the above works without a context. Although I admit the example is far-sought and feels counter intuitive. 

What also bothers me a bit is that all the checks are currently in `MethodResolverImpl`. Here we do allow explicitly same-type mapping methods to override the "direct assignable" types. The check feels a bit out of place here. At the other hand: we don't have any knowledge from the calling side in `MethodResolverImpl` (so it does not it's a constant).

An argument for this solution is that it used to work. Apparently we "broke" something. 

@filiphr WDYT?

Reply:
This is a tricky one @sjaakd. It broke due to the fixes we did in #1714. If there are qualifiers then we do not use the direct assignment.

Your example is entirely OK @sjaakd, and it is even valid without the `@Context`. Currently, we will pick up the `defaultToString` method. With this PR that is not the case.

A more simpler example is something like:


```
@Mapper
public interface MyMapper {

    @Mapping(target = "name", source = "firstName", defaultValue = "unknown", qualifiedByName = "uppercase")
    Customer map(CustomerDto dto);

    @Named("uppercase")
    default String uppercase(String value) {
        return value == null ? null : value.toUpperCase(Locale.ROOT);
    }
}
```

Before this PR we would use the `uppercase` method for both, which IMO is correct. With this PR that is no longer the case.

I am more keen on not providing fix for this and telling the user what the actual problem is, since if we start going down this path there will be even more problems. We also have `defaultValueExpression` for using things literally.

Reply:
Concerning that uppercase example, wouldn't it be weird that the `defaultValue` listed there is not in uppercase itself? The user specifies a `defaultValue`, but instead of that one being used it gets adjusted by the mapping method for the non-default values.

Also I find it weird that the behavior of `defaultValue` and `defaultExpression` differs. The only thing I would expect to be different with these is that one uses a `String` or `Primitive Data Type` input as a default value, while the other can handle all the other types using methods. Using your example of uppercase, wouldn't it be more logical to suggest here to use either `defaultValue="UNKNOWN"` or `defaultExpression=java(uppercase("unknown"))`?

On the other hand the example that @sjaakd gave is one I did not think of. That situation would make the `defaultValue` usage the way I thought of it break. Currently I end up with at least 2 other options to solve this situation:

1. Maybe an option is to leave it up to the user to decide how to handle these situations. We could introduce a setting in the `@MappingControl` to handle this. [The changes would look something like this.](https://github.com/Zegveld/mapstruct/compare/feature/2636...Zegveld:feature/2636_alternative) That alternative solution is not finished yet, it needs more situations for testing it. Including those with an `@Context` annotation and perhaps quite a bit more work.
2. Update the Reference Guide to mention that `defaultValue` and `constant` are also mapped through the qualified mappings if possible, and that if this is not wanted they should use the `defaultExpression` variant instead.

Perhaps the 2nd option is the better one, that we document it and if people want to influence it we can always add the 1st option upon request.

Reply:
My example was extremely simplified. The implementation logic can be anything, perhaps you have some kind of a mapping by key and you need to get an id from the DB, similar to what Sjaak showed, but not with a `@Context`.

I think that option 2 is the best here. Update the Reference Guide and also add something to the javadoc of those methods as well. Perhaps we can also add some tests for this, to verify that everything works like we expect it to (like it is now), since as we can see there are no tests failing currently 😄

Reply:
> Your example is entirely OK @sjaakd, and it is even valid without the `@Context`. Currently, we will pick up the `defaultToString` method. With this PR that is not the case.

The `@Context` was intended to show that even a defaultValue might be context dependent.

Reply:
> Also I find it weird that the behavior of `defaultValue` and `defaultExpression` differs. The only thing I would expect to be different with these is that one uses a `String` or `Primitive Data Type` input as a default value, while the other can handle all the other types using methods.

Some history: They are in essence different because the `expression` we introduced long time ago with the idea the user would have a means to solve all kinds of shortcomings that we did not have an answer for (yet) 😄 . The reason that it has a syntax like `java(...)` is that we anticipated other syntaxes, like `el(..)` for expression language.  I even thought about using antlr and used the jsr of el to generate an AST for generarting code. But I deemed it to cumbersome. 

So in the end we decided to do no interpretation at all. In contrast: a `constant` was supposed to be a `String` from the start. Later we added the relatively new option to interpret java literals: constants for the primitive types. Like `5d`, `3L` etc.

The default variants evolved somewhere in the middle.